server/src/main/java/org/elasticsearch/cluster/metadata/OperatorHandlerMetadata.java

```diff
@@ -0,0 +1,144 @@
+ */
+public record OperatorHandlerMetadata(String name, Set<String> keys)
+    implements
+        SimpleDiffable<OperatorHandlerMetadata>,
+        ToXContentFragment {
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(name);
+        out.writeCollection(keys, StreamOutput::writeString);
+    }
+
+    /**
+     * Reads an {@link OperatorHandlerMetadata} from a {@link StreamInput}
+     * @param in the {@link StreamInput} to read from
+     * @return {@link OperatorHandlerMetadata}
+     * @throws IOException
+     */
+    public static OperatorHandlerMetadata readFrom(StreamInput in) throws IOException {
+        Builder builder = new Builder(in.readString()).keys(in.readSet(StreamInput::readString));
+        return builder.build();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        Builder.toXContent(this, builder);
+        return builder;
+    }
+
+    /**
+     * Reads an {@link OperatorHandlerMetadata} {@link Diff} from {@link StreamInput}
+     * @param in the {@link StreamInput} to read the diff from
+     * @return a {@link Diff} of {@link OperatorHandlerMetadata}
+     * @throws IOException
+     */
+    public static Diff<OperatorHandlerMetadata> readDiffFrom(StreamInput in) throws IOException {
+        return SimpleDiffable.readDiffFrom(OperatorHandlerMetadata::readFrom, in);
+    }
+
+    /**
+     * Builder class for {@link OperatorHandlerMetadata}
+     */
+    public static class Builder {
+        private static final String HANDLER_KEYS = "keys";
+
+        private final String name;
+        private Set<String> keys;
+
+        /**
+         * Default constructor, the handler name is a required parameter
+         * @param name the handler name
+         */
+        public Builder(String name) {
+            this.name = name;
+            this.keys = new HashSet<>();
+        }
+
+        /**
+         * Set the cluster state keys recorded/written by this handler in the cluster state
+         * @param keys a set of String keys of the state that's written by this handler
+         * @return {@link Builder}
+         */
+        public Builder keys(Set<String> keys) {
+            this.keys = keys;
+            return this;
+        }
+
+        /**
+         * Builds the {@link OperatorHandlerMetadata}
+         * @return {@link OperatorHandlerMetadata}
+         */
+        public OperatorHandlerMetadata build() {
+            return new OperatorHandlerMetadata(name, Collections.unmodifiableSet(keys));
+        }
+
+        /**
+         * Serializes an {@link OperatorHandlerMetadata} to xContent
+         *
+         * @param metadata {@link OperatorHandlerMetadata}
+         * @param builder {@link XContentBuilder}
+         */
+        public static void toXContent(OperatorHandlerMetadata metadata, XContentBuilder builder) throws IOException {
```

Comment:
Why is toXContent needed on both the builder (as a static that takes the record?) and on the record itself?

Reply:
I'll move it to the record only, similar to other places, I'm sharing the string constants for to and from xcontent.